open_spiel/games/oh_hell.cc

```diff
@@ -0,0 +1,589 @@
+    /*provides_observation_string=*/false,
+    /*provides_observation_tensor=*/false,
+    /*parameter_specification=*/
+    {
+        {"players", GameParameter(kMinNumPlayers)},
+        {"num_suits", GameParameter(kMaxNumSuits)},
+        {"num_cards_per_suit", GameParameter(kMaxNumCardsPerSuit)},
+        // number of tricks in the game, must be between 1 and
+        // (num_suits * num_cards_per_suit - 1) / num_players,
+        // default is to choose randomly in the legal range every game
+        {"num_tricks_fixed", GameParameter(kRandomNumTricks)},
+    }};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new OhHellGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+}  // namespace
+
+OhHellGame::OhHellGame(const GameParameters& params)
+    : Game(kGameType, params),
+      num_players_(ParameterValue<int>("players")),
+      deck_props_(ParameterValue<int>("num_suits"),
+                  ParameterValue<int>("num_cards_per_suit")),
+      num_tricks_fixed_(ParameterValue<int>("num_tricks_fixed")) {
+  SPIEL_CHECK_TRUE(num_players_ >= kMinNumPlayers &&
+                   num_players_ <= kMaxNumPlayers);
+  SPIEL_CHECK_TRUE(deck_props_.NumSuits() >= kMinNumSuits &&
+                   deck_props_.NumSuits() <= kMaxNumSuits);
+  SPIEL_CHECK_TRUE(deck_props_.NumCardsPerSuit() >= kMinNumCardsPerSuit &&
+                   deck_props_.NumCardsPerSuit() <= kMaxNumCardsPerSuit);
+  // need at least num_players + 1 cards
+  SPIEL_CHECK_TRUE(num_players_ <= deck_props_.NumCards() - kNumTrumpDeal);
+  SPIEL_CHECK_TRUE(num_tricks_fixed_ == kRandomNumTricks ||
+                   (num_tricks_fixed_ >= kMinNumTricks &&
+                   num_tricks_fixed_ <= MaxNumTricks()));
+}
+
+std::vector<int> OhHellGame::InformationStateTensorShape() const {
+  // initial chance actions (incl trump dealing)
+  int len = MaxNumTricks() + num_players_ + deck_props_.NumCards();
+  // initial hand and current hand
+  len += 2 * deck_props_.NumCards();
+  // bids, legal range is [no bid, 0, 1, ..., max legal bid]
+  len += num_players_ * (MaxNumTricks() + 2);
+  // tricks won so far
+  len += MaxNumTricks() * num_players_;
+  // tricks
+  len += MaxNumTricks() * (2 * num_players_ - 1) * deck_props_.NumCards();
+  return {len};
+}
+
+OhHellState::OhHellState(std::shared_ptr<const Game> game, int num_players,
+                         DeckProperties deck_props, int num_tricks_fixed)
+    : State(game),
+      num_players_(num_players),
+      deck_props_(deck_props),
+      num_tricks_fixed_(num_tricks_fixed) {
+  bids_.resize(num_players_);
+  // need to differentiate between no bid and a bid of 0
+  std::fill(bids_.begin(), bids_.end(), kInvalidBid);
+  num_tricks_won_.resize(num_players_);
+  returns_.resize(num_players_);
+  holder_.resize(deck_props_.NumCards());
+  initial_deal_.resize(deck_props_.NumCards());
+}
+
+std::string OhHellState::ToString() const {
```

Comment:
I think it would be helpful to add (a string representation) of the the current game phase ("ChooseNumTricks", "Dealer", ...) at the beginning of the state string